lang/LangPrimSource/SC_ComPort.h

```diff
@@ -20,103 +20,122 @@
 
 const int kTextBufSize = 65536;
 
-class SC_UdpInPort {
-    int mPortNum;
-    boost::array<char, kTextBufSize> recvBuffer;
+enum class HandlerType { OSC, Raw };
 
-    boost::asio::ip::udp::endpoint remoteEndpoint;
+using HandleDataFunc = std::function<void(std::unique_ptr<char[]>, size_t)>;
 
-    void handleReceivedUDP(const boost::system::error_code& error, std::size_t bytes_transferred);
+//////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-    void startReceiveUDP();
+namespace Detail {
 
+class TCPConnection : public std::enable_shared_from_this<TCPConnection>, private boost::noncopyable {
 public:
-    boost::asio::ip::udp::socket udpSocket;
+    using pointer = std::shared_ptr<TCPConnection>;
 
-    int RealPortNum() const { return mPortNum; }
-    boost::asio::ip::udp::socket& Socket() { return udpSocket; }
+    TCPConnection(boost::asio::io_service& ioService, int portNum, HandlerType);
 
-    SC_UdpInPort(int inPortNum, int portsToCheck = 10);
-    ~SC_UdpInPort();
-};
+    void start();
+    auto& getSocket() { return mSocket; }
 
-class SC_UdpCustomInPort : public SC_UdpInPort {
-public:
-    SC_UdpCustomInPort(int inPortNum);
-    ~SC_UdpCustomInPort();
+private:
+    void handleLengthReceived(const boost::system::error_code& error, size_t bytes_transferred);
+    void handleMsgReceived(const boost::system::error_code& error, size_t bytes_transferred);
+    void initHandler(HandlerType);
+
+    HandleDataFunc mHandleFunc;
+    boost::asio::ip::tcp::socket mSocket;
+    int32 mOSCMsgLength;
+    std::unique_ptr<char[]> mData;
+    const int mPortNum;
 };
+}
 
 //////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-class SC_TcpConnection : public boost::enable_shared_from_this<SC_TcpConnection> {
-public:
-    typedef boost::shared_ptr<SC_TcpConnection> pointer;
-    boost::asio::ip::tcp::socket socket;
+namespace InPort {
 
-    SC_TcpConnection(boost::asio::io_service& ioService, class SC_TcpInPort* parent):
-        socket(ioService),
-        mParent(parent) {}
+class UDP : private boost::noncopyable {
+public:
+    UDP(int inPortNum, HandlerType, int portsToCheck = 10);
+    ~UDP() = default;
 
-    void start();
+    auto RealPortNum() const { return mPortNum; }
+    auto& getSocket() { return mUdpSocket; }
 
 private:
-    int32 OSCMsgLength;
-    char* data;
-    class SC_TcpInPort* mParent;
+    void initHandler(HandlerType type);
 
-    void handleLengthReceived(const boost::system::error_code& error, size_t bytes_transferred);
+    void handleReceivedUDP(const boost::system::error_code& error, std::size_t bytes_transferred);
+    void startReceiveUDP();
 
-    void handleMsgReceived(const boost::system::error_code& error, size_t bytes_transferred);
+    int mPortNum;
+    HandleDataFunc mHandleFunc;
+    boost::array<char, kTextBufSize> mRecvBuffer;
```

Comment:
is `boost:array` the same as `std::array`?

Reply:
This should be `std::array`. Resolved.